hail/src/main/scala/is/hail/expr/ir/functions/UtilFunctions.scala

```diff
@@ -125,39 +125,41 @@ object UtilFunctions extends RegistryFunctions {
   def registerAll() {
     val thisClass = getClass
 
-    registerCode("valuesSimilar", tv("T"), tv("U"), TFloat64, TBoolean, TBoolean, null) {
+    registerCode("valuesSimilar", tv("T"), tv("U"), TFloat64, TBoolean, TBoolean, {
+      case(_: PType, _: PType, _: PType, _: PType) => PBoolean()
+    }) {
       case (er, rt, (lT, l), (rT, r), (tolT, tolerance), (absT, absolute)) =>
         assert(lT.virtualType == rT.virtualType)
         val lb = boxArg(er, lT)(l)
         val rb = boxArg(er, rT)(r)
         er.mb.getType(lT.virtualType).invoke[Any, Any, Double, Boolean, Boolean]("valuesSimilar", lb, rb, tolerance, absolute)
     }
 
-    registerCode[Int]("triangle", TInt32, TInt32, null) { case (_, rt, (nT, n: Code[Int])) =>
+    registerCode[Int]("triangle", TInt32, TInt32, (n: PType) => n) { case (_, rt, (nT, n: Code[Int])) =>
```

Comment:
not entirely certain of this, but I thought that the operations on the Code value corresponding to nT should remain of type nT.

Reply:
this seems slightly different from invoking a Scala operation on the JVM values generated from code values as we discussed yesterday

Reply:
This seems basically the same, I think it should be PInt32()